src/cast_core/src/actors/CrudActor.cpp

```diff
@@ -0,0 +1,615 @@
+        auto maxStaleness = rhs.max_staleness();
+        if (maxStaleness) {
+            node["MaxStalenessSeconds"] = genny::TimeSpec(*maxStaleness).count();
+        }
+        return node;
+    }
+
+    static bool decode(const Node& node, ReadPreference& rhs) {
+        if (!node.IsMap()) {
+            return false;
+        }
+        if (!node["ReadMode"]) {
+            // readPreference must have a read mode specified.
+            return false;
+        }
+        auto readMode = node["ReadMode"].as<std::string>();
+        rhs = mongocxx::read_preference{};
+        if (readMode == "primary") {
+            rhs.mode(ReadMode::k_primary);
+        } else if (readMode == "primaryPreferred") {
+            rhs.mode(ReadMode::k_primary_preferred);
+        } else if (readMode == "secondary") {
+            rhs.mode(ReadMode::k_secondary);
+        } else if (readMode == "secondaryPreferred") {
+            rhs.mode(ReadMode::k_secondary_preferred);
+        } else if (readMode == "nearest") {
+            rhs.mode(ReadMode::k_nearest);
+        } else {
+            return false;
+        }
+        if (node["MaxStalenessSeconds"]) {
+            auto maxStaleness = node["MaxStalenessSeconds"].as<int>();
+            rhs.max_staleness(std::chrono::seconds(maxStaleness));
+        }
+        return true;
+    }
+};
+
+template <>
+struct convert<mongocxx::write_concern> {
+    using WriteConcern = mongocxx::write_concern;
+    static Node encode(const WriteConcern& rhs) {
+        Node node;
+        auto timeout = rhs.timeout();
+        auto timeoutCount = genny::TimeSpec{timeout}.count();
+        node["TimeoutMillis"] = timeoutCount;
+
+        auto journal = rhs.journal();
+        node["Journal"] = journal;
+
+        auto ackLevel = rhs.acknowledge_level();
+        if (ackLevel == WriteConcern::level::k_majority) {
+            node["Level"] = "majority";
+        } else if (ackLevel == WriteConcern::level::k_acknowledged) {
+            auto numNodes = rhs.nodes();
+            if (numNodes) {
+                node["Level"] = *numNodes;
+            }
+        }
+        return node;
+    }
+
+    static bool decode(const Node& node, WriteConcern& rhs) {
+        if (!node.IsMap()) {
+            return false;
+        }
+        if (!node["Level"]) {
+            // writeConcern must specify the write concern level.
+            return false;
+        }
+        auto level = node["Level"].as<std::string>();
+        bool isNum = level.find_first_not_of("0123456789") == std::string::npos;
+        if (isNum) {
+            auto levelNum = node["Level"].as<int>();
+            rhs.nodes(levelNum);
+        } else if (level == "majority") {
+            rhs.majority(std::chrono::milliseconds{0});
+        } else {
+            // writeConcern level must be of valid integer or 'majority'.
+            return false;
+        }
+        if (node["TimeoutMillis"]) {
+            auto timeout = node["TimeoutMillis"].as<int>();
```

Comment:
Per genny's conventions of having time-based units be a first-class concept, you'll want to have this be a `TimeSpec` rather than an `int`. This can be achieved by delegating to `TimeSpec`'s decoder.

```cpp
if (node["Timeout"]) {
    auto timeout = node["Timeout"].as<genny::TimeSpec>();
    rhs.timeout(std::chrono::milliseconds{timeout});
}
```

Similar for the `encode()` function  you'll want to do `node["Timeout"] = genny::TimeSpec{rhs.timeout()}`.